KAFKA-17645: KIP-1052: Enable warmup in producer performance test

matt-welch opened the pull request:
In order to better analyze steady-state performance of Kafka, this PR
enables a warmup in the Producer Performance test. The warmup duration
is specified as a number of records that are a subset of the total
numRecords. If warmup records is greater than 0, the warmup is
represented by a second Stats object which holds warmup results. Once
warmup records have been exhausted, the test switches to using the
existing Stats object. At end of test, if warmup was enabled, the
summary of the whole test (warump + steady state) is printed followed by
the summary of the steady-state portion of the test.  If no warmup is
used, summary prints don't change from existing behavior. This
contribution is an original work and is licensed to the Kafka project
under the Apache license

Testing strategy comprises new Java unit tests added to
ProducerPerformanceTests.java.

### Committer Checklist (excluded from commit message)
- [X] Verify design and implementation
- [X] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

Reviewers: Kirk True <kirk@kirktrue.pro>, Federico Valeri
 <fedevaleri@gmail.com>, Chia-Ping Tsai <chia7712@gmail.com>

matt-welch committed:
Refactored for trunk

This commit should enable a test PR against the trunk

matt-welch committed:
Refactor for readability

Now using an overallStats object with new constructor to contin the
combined stats at end of test and a single printTotal method.

matt-welch committed:
:bugfix: Remove redundant declaration of overallStats

matt-welch committed:
Remove overallStats

Refactor to remove declaration of overallStats and limit its scope to
only the location where printTotal is called.
Set Stats combine-constructor as package-private.
Remove commented code.

matt-welch committed:
Merge branch 'trunk' into KIP-1052-test-PR

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -75,7 +75,13 @@ void start(String[] args) throws IOException {
             ProducerRecord<byte[], byte[]> record;
-            stats = new Stats(config.numRecords, 5000);
+            if (config.warmupRecords > 0) {
+                // TODO: Keep this message? Maybe unnecessary
+                System.out.println("Warmup first " + config.warmupRecords + " records. Steady-state results will print after the complete-test summary.");
```
Using the tool I changed my mind about this. It's useful to know that the tool is running the warmup, but we would also need a similar log for the steady state phase, otherwise you wouldn't be able to distinguish if the periodic log is about warmup or steady state.

matt-welch replied:
Are you suggesting we have an additional print message that happens during the switchover to steady state?

fvaleri replied:
Exactly. I think it could be useful.

matt-welch replied:
I like this to explicitly inform the user that they're in steady state.  
The wording is up for debate, but I've added a basic message "Beginning steady state" to the record() method to indicate to the user that following prints are steady state.  In testing, I've observed that you might not see window prints for different number of records, depending on how you set your warmup records, relative the size of the print window.  This is similar to behavior in the existing test where the final window gets included in the final summary print.

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -75,7 +75,13 @@ void start(String[] args) throws IOException {
+            if (config.warmupRecords > 0) {
+                // TODO: Keep this message? Maybe unnecessary
+                System.out.println("Warmup first " + config.warmupRecords + " records. Steady-state results will print after the complete-test summary.");
+                this.warmupStats = new Stats(config.warmupRecords, 5000);
+            } else {
+                stats = new Stats(config.numRecords, 5000);
```
We probably want to keep them aligned, so I would create the `DEFAULT_REPORTING_INTERVAL_MS` constant.

matt-welch replied:
Agreed.  Added a constant in next push

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -94,8 +100,18 @@ void start(String[] args) throws IOException {
+                        stats = new Stats(config.numRecords - config.warmupRecords, 5000, config.warmupRecords);
+                        cb = new PerfCallback(sendStartMs, payload.length, stats);
+                    }
+                    producer.send(record, cb);
```
We can move this to line 115 and remove the duplication in line 113.

matt-welch replied:
Agreed.  Fixed in next push

fvaleri commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -428,8 +484,13 @@ public void printTotal() {
+            }
+            System.out.printf("%d%s records sent, %f records/sec (%.2f MB/sec), %.2f ms avg latency, %.2f ms max latency, %d ms 50th, %d ms 95th, %d ms 99th, %d ms 99.9th.%n",
                               count,
+                              state,
```
Can't you just use the ternary operator here?

matt-welch replied:
Agreed.  Fixed in latest push

fvaleri reviewed (commented):
Hi @matt-welch, thanks for working on this.

I didn't have time to look further into this, but new results don't make much sense to me:

```sh
$ bin/kafka-producer-perf-test.sh --topic my-topic --record-size 1024 --num-records 1000000 --throughput 10000 --warmup-records 60000 --producer-props bootstrap.servers=:9092 linger.ms=0 batch.size=16384
...
1000000 records sent, 9998.000400 records/sec (0.00 MB/sec), 44181.96 ms avg latency, 94001.00 ms max latency, 266 ms 50th, 266 ms 95th, 266 ms 99th, 266 ms 99.9th.
940000 steady state records sent, 9997.553790 records/sec (9.76 MB/sec), 47001.95 ms avg latency, 94001.00 ms max latency, 47002 ms 50th, 89301 ms 95th, 93062 ms 99th, 93908 ms 99.9th.
```

As you can see, we have 0.00 MB/sec in total results, and very high latencies in steady-state result. This is the output without your changes:

```sh
1000000 records sent, 9998.0 records/sec (9.76 MB/sec), 0.76 ms avg latency, 255.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 38 ms 99.9th.
```

Another problem I found is the following:

```sh
$ bin/kafka-producer-perf-test.sh --topic my-topic --payload-file LICENSE --num-records -10 --throughput 50000 --producer-props bootstrap.servers=:9092 linger.ms=0 batch.size=16384
usage: producer-performance [-h] --topic TOPIC --num-records NUM-RECORDS [--payload-delimiter PAYLOAD-DELIMITER] --throughput THROUGHPUT [--producer-props PROP-NAME=PROP-VALUE [PROP-NAME=PROP-VALUE ...]] [--producer.config CONFIG-FILE] [--print-metrics] [--transactional-id TRANSACTIONAL-ID]
                            [--transaction-duration-ms TRANSACTION-DURATION] [--warmup-records WARMUP-RECORDS] (--record-size RECORD-SIZE | --payload-file PAYLOAD-FILE | --payload-monotonic)
producer-performance: error: The value for --warmup-records must be strictly fewer than the number of records in the test, --num-records.
```

The check is fine, but it should be skipped if the warmup flag is not used. I also think we should add a non-negative check for `numRecords` in `ConfigPostProcessor`. I opened [another PR](https://github.com/apache/kafka/pull/17348) with similar improvements.

matt-welch committed:
:bugfix: Properly create callback during switch

- Fix callbacks after switch to steady state which were not being
  properly created.
- Add constant for DEFAULT_REPORTING_INTERVAL_MS
- Remove redundant producer.send calls
- Add ternerary operator in state assignment
- Add check for non-positive numRecords

matt-welch commented:
I was able to reproduce and fix the performance issue.  I had failed to create the record callback for messages in steady state :( It should be fixed in the latest push, but I still think the logic in that conditional can be improved.  

re: checking warmup records before num records, I've added an exception for when numRecords <= 0.

FYI, I will be out for the next week, but thanks so much for looking at this patch!!

matt-welch committed:
Merge branch 'trunk' into KIP-1052-PR

matt-welch committed:
Fix redundant numRecords declaration

matt-welch committed:
Fix erroneous extra bracket

matt-welch committed:
Fix ternary operator in printTotal

matt-welch committed:
Fix selection in ternary operator for proper label

fvaleri commented:
Thanks for the latest changes. Results look much better now:

```sh
2000000 records sent, 9998.350272 records/sec (0.00 MB/sec), 0.58 ms avg latency, 417.00 ms max latency, 417 ms 50th, 417 ms 95th, 417 ms 99th, 417 ms 99.9th.
1500000 steady state records sent, 9997.733847 records/sec (0.95 MB/sec), 0.52 ms avg latency, 9.00 ms max latency, 1 ms 50th, 1 ms 95th, 1 ms 99th, 2 ms 99.9th.
```

There is still an issue with the overall throughput (0.00 MB/sec) though. I also see some checkstyle errors after the rebase.

> FYI, I will be out for the next week, but thanks so much for looking at this patch!

No rush, take your time.

matt-welch committed:
Fix concatenating stats constructor

- bytes are now properly accoutned for.
- Refactor to use boolean isSteadyState over warmupRecords to track
  if a Stats object is of steady state

fvaleri commented:
Hi @matt-welch, there is one checkstyle error and you also probably need to rebase.

matt-welch committed:
Fix checkstyle error and print

matt-welch committed:
Merge branch 'trunk' into KIP-1052-PR

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -75,7 +76,13 @@ void start(String[] args) throws IOException {
             ProducerRecord<byte[], byte[]> record;
-            stats = new Stats(config.numRecords, 5000);
+            if (config.warmupRecords > 0) {
+                // TODO: Keep this message? Maybe unnecessary
```
yes, please the unnecessary log

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -114,6 +133,10 @@ record = new ProducerRecord<>(config.topicName, payload);
 
+                /* print warmup stats if relevant */
+                if (warmupStats != null) {
+                    new Stats(warmupStats, stats).printTotal();
```
Maybe we should have a state for 'all' and another state for 'steady-only'. With this approach, we can avoid this kind of magic merging.

matt-welch replied:
Thanks for the review @chia7712!
Just to be clear, are you suggesting that, instead of using a warmup stats followed by a steady state stats object, we should use a whole-test stats object in addition to a steady-state stats object?
I don't see any serious issues in the implementation, but I the producer callback will need refactoring so it writes into two stats objects once steady state has triggered. In the initial design, I was trying to avoid this concurrent need for both stats objects and do the array-merging at the end, but I will give it a try to see how it looks.

fvaleri replied:
@matt-welch did you tried the suggested approach?

matt-welch replied:
Sorry for the delay.  I've refactored the stats objects as suggested in my latest commit.

chia7712 reviewed (commented):
@matt-welch thanks for this implementation!

fvaleri commented:
@matt-welch sorry for the delay. Latest changes LGTM. Please, address the comments from @chia7712 and then I think we are good. Thanks.

matt-welch committed:
Refactor for redundant stats objects

Stats objects now represent the whole test with
'stats' and the steady state portion of the test
with 'steadyStateStats'. Remove merging of
Stats latency arrays at end of test.

matt-welch commented:
Hi @chia7712 @fvaleri .  My latest commit refactors the use of a warmupStats object and a main stats object in favor of having one stats that covers the whole test and a second object, steadyStateStats, that will be used during steady state operation.  
In the case of a steady state record, I've added a new callback that will write into both data objects during steady state.
This new approach seems like a simpler, cleaner design and is much easier to understand what is happening. I've done some testing on both designs and they appear to perform similarly. My only reservation is that the producer will now have a larger memory requirement due to the redundant recording of latency data into the two objects.  Let me know what you think.

fvaleri reviewed (approved):
LGTM. Nice feature. Thanks @matt-welch.

@chia7712 call for review.

chia7712 commented:
@matt-welch thanks for your update. I will take a look later!

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -361,6 +400,26 @@ public Stats(long numRecords, int reportingInterval) {
+            this.isSteadyState = isSteadyState;
+        }
+
+        Stats(Stats first, Stats second) {
```
Could you please remove this unused constructor?

matt-welch replied:
Done

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -94,7 +98,18 @@ void start(String[] args) throws IOException {
+                        cb = new PerfCallback(sendStartMs, payload.length, stats);
+                    } else {
+                        if (i == config.warmupRecords) {
+                            steadyStateStats = new Stats(config.numRecords - config.warmupRecords, DEFAULT_REPORTING_INTERVAL_MS, config.warmupRecords > 0);
```
`config.warmupRecords > 0` can be replaced by `true`

matt-welch replied:
Can you please clarify what you mean?  That conditional is required when warmup is enabled.

chia7712 replied:
Apologies for the unclear comment. Please see the following code.
```
if (config.warmupRecords > 0) {
                    if (i < config.warmupRecords) {
                        cb = new PerfCallback(sendStartMs, payload.length, stats);
                    } else {
                        if (i == config.warmupRecords) {
                            steadyStateStats = new Stats(config.numRecords - config.warmupRecords, DEFAULT_REPORTING_INTERVAL_MS, config.warmupRecords > 0);
                        }
                        cb = new PerfCallback(sendStartMs, payload.length, stats, steadyStateStats);
                    }
                }
```

In this case, config.warmupRecords > 0 is always true.

matt-welch replied:
Apologies if I'm still misunderstanding your query, but the conditional "config.warmupRecords > 0" will only evaluate to true when the argument '--warmup-records N' is invoked on the command line and N is greater than zero. The value for config.warmupRecords defaults to '0' otherwise. The if-clause on that line is the path used when a warmup is invoked. The else-clause will be used when no warmup has been requested. This else clause represents the same code path used in previous versions of the test with no warmup.

chia7712 replied:
Apologies for the unclear description. I have attached a screenshot to illustrate the warning from the IDE.

![Screenshot From 2024-12-18 19-08-16](https://github.com/user-attachments/assets/bc3da099-caf8-4cea-87e4-d3346e85c665)

we can make IDE happy by a little change :)

matt-welch replied:
Hi @chia7712, to address your concern, I have temporarily added a DEBUG message that prints the value of config.warmupRecords and the value of the conditional (config.warmupRecords > 0) before the test.  Some sample output of running the test with various values of warmupRecords follows.  Note that, when negative values of warmupRecords are supplied on the CLI, we set config.warmupRecords to 0.

```
# warmupRecords = -1
./kafka-producer-perf-test.sh --producer.config ../config/producer.properties --topic warmup--1-records-test --num-records 500000 --record-size 1024 --producer-props bootstrap.servers=127.0.0.1:9092 acks=all compression.type=none batch.size=16384 linger.ms=0 --throughput 10000 --warmup-records -1
DEBUG: config.warmupRecords=0, (config.warmupRecords > 0)=false
49987 records sent, 9997.4 records/sec (9.76 MB/sec), 18.8 ms avg latency, 290.0 ms max latency.
50005 records sent, 10001.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 9.0 ms max latency.
50019 records sent, 10001.8 records/sec (9.77 MB/sec), 0.7 ms avg latency, 9.0 ms max latency.
50005 records sent, 10001.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50006 records sent, 9999.2 records/sec (9.76 MB/sec), 0.7 ms avg latency, 3.0 ms max latency.
49930 records sent, 9960.1 records/sec (9.73 MB/sec), 0.7 ms avg latency, 22.0 ms max latency.
50229 records sent, 10045.8 records/sec (9.81 MB/sec), 0.7 ms avg latency, 23.0 ms max latency.
50015 records sent, 10003.0 records/sec (9.77 MB/sec), 0.6 ms avg latency, 2.0 ms max latency.
49995 records sent, 9997.0 records/sec (9.76 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
500000 records sent, 9996.801024 records/sec (9.76 MB/sec), 2.49 ms avg latency, 290.00 ms max latency, 1 ms 50th, 1 ms 95th, 82 ms 99th, 219 ms 99.9th.

# warmupRecords = 0
./kafka-producer-perf-test.sh --producer.config ../config/producer.properties --topic warmup-0-records-test --num-records 500000 --record-size 1024 --producer-props bootstrap.servers=127.0.0.1:9092 acks=all compression.type=none batch.size=16384 linger.ms=0 --throughput 10000 --warmup-records 0
DEBUG: config.warmupRecords=0, (config.warmupRecords > 0)=false
49982 records sent, 9996.4 records/sec (9.76 MB/sec), 2.3 ms avg latency, 177.0 ms max latency.
50010 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 4.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 12.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 3.0 ms max latency.
50020 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50008 records sent, 10001.6 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
500000 records sent, 9997.200784 records/sec (9.76 MB/sec), 0.83 ms avg latency, 177.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 44 ms 99.9th.

# warmupRecords = 1
./kafka-producer-perf-test.sh --producer.config ../config/producer.properties --topic warmup-1-records-test --num-records 500000 --record-size 1024 --producer-props bootstrap.servers=127.0.0.1:9092 acks=all compression.type=none batch.size=16384 linger.ms=0 --throughput 10000 --warmup-records 1
DEBUG: config.warmupRecords=1, (config.warmupRecords > 0)=true
Warmup first 1 records. Steady state results will print after the complete test summary.
Beginning steady state.
51651 records sent, 10328.1 records/sec (10.09 MB/sec), 2.7 ms avg latency, 63.0 ms max latency.
50015 records sent, 10003.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 3.0 ms max latency.
49995 records sent, 9999.0 records/sec (9.76 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50013 records sent, 10002.6 records/sec (9.77 MB/sec), 0.7 ms avg latency, 3.0 ms max latency.
50016 records sent, 10001.2 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50011 records sent, 10000.2 records/sec (9.77 MB/sec), 0.7 ms avg latency, 19.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50010 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
500000 records sent, 9997.800484 records/sec (9.76 MB/sec), 0.87 ms avg latency, 177.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 54 ms 99.9th.
499999 steady state records sent, 10030.070211 records/sec (9.79 MB/sec), 0.87 ms avg latency, 63.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 54 ms 99.9th.

# warmupRecords = 10000
./kafka-producer-perf-test.sh --producer.config ../config/producer.properties --topic warmup-10000-records-test --num-records 500000 --record-size 1024 --producer-props bootstrap.servers=127.0.0.1:9092 acks=all compression.type=none batch.size=16384 linger.ms=0 --throughput 10000 --warmup-records 10000
DEBUG: config.warmupRecords=10000, (config.warmupRecords > 0)=true
Warmup first 10000 records. Steady state results will print after the complete test summary.
Beginning steady state.
49982 records sent, 9996.4 records/sec (9.76 MB/sec), 0.7 ms avg latency, 4.0 ms max latency.
50010 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 3.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50020 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 7.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 4.0 ms max latency.
50009 records sent, 10001.8 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50001 records sent, 9998.2 records/sec (9.76 MB/sec), 0.6 ms avg latency, 2.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.6 ms avg latency, 2.0 ms max latency.
500000 records sent, 9996.201443 records/sec (9.76 MB/sec), 0.83 ms avg latency, 178.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 46 ms 99.9th.
490000 steady state records sent, 9995.308325 records/sec (9.76 MB/sec), 0.66 ms avg latency, 7.00 ms max latency, 1 ms 50th, 1 ms 95th, 1 ms 99th, 2 ms 99.9th.

# warmupRecords = 50000
./kafka-producer-perf-test.sh --producer.config ../config/producer.properties --topic warmup-50000-records-test --num-records 500000 --record-size 1024 --producer-props bootstrap.servers=127.0.0.1:9092 acks=all compression.type=none batch.size=16384 linger.ms=0 --throughput 10000 --warmup-records 50000
DEBUG: config.warmupRecords=50000, (config.warmupRecords > 0)=true
Warmup first 50000 records. Steady state results will print after the complete test summary.
Beginning steady state.
49991 records sent, 9998.2 records/sec (9.76 MB/sec), 0.6 ms avg latency, 3.0 ms max latency.
50011 records sent, 10002.2 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 25.0 ms max latency.
50010 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 12.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 4.0 ms max latency.
50020 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.6 ms avg latency, 2.0 ms max latency.
50020 records sent, 10002.0 records/sec (9.77 MB/sec), 0.6 ms avg latency, 2.0 ms max latency.
500000 records sent, 9997.600576 records/sec (9.76 MB/sec), 0.81 ms avg latency, 172.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 40 ms 99.9th.
450000 steady state records sent, 9996.001599 records/sec (9.76 MB/sec), 0.66 ms avg latency, 25.00 ms max latency, 1 ms 50th, 1 ms 95th, 1 ms 99th, 4 ms 99.9th.

# warmupRecords = 100000
./kafka-producer-perf-test.sh --producer.config ../config/producer.properties --topic warmup-100000-records-test --num-records 500000 --record-size 1024 --producer-props bootstrap.servers=127.0.0.1:9092 acks=all compression.type=none batch.size=16384 linger.ms=0 --throughput 10000 --warmup-records 100000
DEBUG: config.warmupRecords=100000, (config.warmupRecords > 0)=true
Warmup first 100000 records. Steady state results will print after the complete test summary.
49994 records sent, 9996.8 records/sec (9.76 MB/sec), 2.2 ms avg latency, 184.0 ms max latency.
Beginning steady state.
49991 records sent, 9994.2 records/sec (9.76 MB/sec), 0.6 ms avg latency, 2.0 ms max latency.
50006 records sent, 10001.2 records/sec (9.77 MB/sec), 0.9 ms avg latency, 31.0 ms max latency.
49995 records sent, 9999.0 records/sec (9.76 MB/sec), 0.6 ms avg latency, 8.0 ms max latency.
50020 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 4.0 ms max latency.
50000 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50010 records sent, 10002.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 2.0 ms max latency.
50010 records sent, 10000.0 records/sec (9.77 MB/sec), 0.7 ms avg latency, 7.0 ms max latency.
500000 records sent, 9997.200784 records/sec (9.76 MB/sec), 0.84 ms avg latency, 184.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 42 ms 99.9th.
400000 steady state records sent, 9995.002499 records/sec (9.76 MB/sec), 0.69 ms avg latency, 31.00 ms max latency, 1 ms 50th, 1 ms 95th, 2 ms 99th, 17 ms 99.9th.
```

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -506,6 +584,9 @@ public ConfigPostProcessor(ArgumentParser parser, String[] args) throws IOExcept
             if (numRecords != null && numRecords <= 0) {
                 throw new ArgumentParserException("--num-records should be greater than zero", parser);
             }
+            if (warmupRecords != null && warmupRecords >= numRecords) {
```
`warmupRecords != null` is unnecessary. Also, could you please change the type of `numRecords` and `warmupRecords` from `Long` to `long`?

matt-welch replied:
Done

chia7712 reviewed (commented):
@matt-welch thanks for this patch. some minor comments are left. Additionally, could you please run this patch and share the result to me? thanks!

Commit:
Merge branch 'trunk' of github.com:apache/kafka into trunk

Commit:
Fix double-printing of windows during steady state

- Add boolean to stats objects to control if they should print or not
- Add DEBUG print to show value of config.warmupRecords

matt-welch committed:
Remove unused constructor, fix types

- Use "long" for numRecords and warmupRecords instead of "Long"
  - Remove "* != null" for above
- Remove unused "combiner" constructor

matt-welch commented:
Hi @chia7712 and @fvaleri.  I've made some updates to this PR in my latest commits including some sample output for review.  Please let me know if there's anything else I can or should do to complete the PR or answer any concerns about it.
Thanks!

chia7712 commented:
@matt-welch sorry for late review. could you please sync the trunk?

kirktrue commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -75,7 +76,13 @@ void start(String[] args) throws IOException {
             ProducerRecord<byte[], byte[]> record;
-            stats = new Stats(config.numRecords, 5000);
+
+            System.out.println("DEBUG: config.warmupRecords=" + config.warmupRecords + ", (config.warmupRecords > 0)=" + (config.warmupRecords > 0));
```
Just for clarity, this `DEBUG` is for development purposes and will be removed before merging?

matt-welch replied:
That's correct. Chia-Ping had asked for some clarification around config.warmupRecords. Once he's satisfied with that output, the line will be removed.

kirktrue commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -94,7 +101,19 @@ void start(String[] args) throws IOException {
                 record = new ProducerRecord<>(config.topicName, payload);
 
                 long sendStartMs = System.currentTimeMillis();
-                cb = new PerfCallback(sendStartMs, payload.length, stats);
```
Any reason not to change this to:

```java
if (config.warmupRecords > 0 && i == config.warmupRecords) {
    steadyStateStats = new Stats(config.numRecords - config.warmupRecords, DEFAULT_REPORTING_INTERVAL_MS, true);
    stats.steadyStateActive = true;
}

cb = new PerfCallback(sendStartMs, payload.length, stats, steadyStateStats);
```

True, `steadyStateStats` will be `null` up until the number of warmup records has elapsed. As I read in `Stats`, passing in a `null` `Stats` object doesn't hurt.

CMIIW.

matt-welch replied:
Thanks for the suggestion!  I think the refactored conditional is much easier to understand. Hopefully nobody minds the passing of null steadyStateStats.  I've refactored this in my latest commit.

kirktrue commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -148,6 +175,7 @@ KafkaProducer<byte[], byte[]> createKafkaProducer(Properties props) {
     Callback cb;
```
I know this is unrelated to these changes, but why are `cb` and `stats` declared at the instance level instead of just created inside `start()`?

matt-welch replied:
Regarding cb and the stats objects, their scopes don't extend beyond the start() method and the callback which gets them passed in.  I tested moving them and found no errors.  Do you think that I should move their declaration in this patch?

kirktrue commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -51,6 +51,7 @@ public class ProducerPerformance {
 
     public static final String DEFAULT_TRANSACTION_ID_PREFIX = "performance-producer-";
     public static final long DEFAULT_TRANSACTION_DURATION_MS = 3000L;
+    public static final int DEFAULT_REPORTING_INTERVAL_MS = 5000;
```
It seems like this constant could be folded into the `Stats` class and removed as a constructor parameter since it's never changed?

matt-welch replied:
Earlier in the PR, Chia-Ping asked to declare this variable here, probably because it seems analogous to the DEFAULT_TRANSACTION_DURATION_MS. I think the intent here is to have this as a parameter than can be modified for more frequent reporting, but making it a final member of the Stats class seems roughly equivalent to me. @kirktrue, @chia7712 where do you think these constants should be declared?

chia7712 replied:
If we don't plan to modify the `reportingInterval` for now, it is fine to remove the `reportingInterval` from `Stats`

matt-welch replied:
My latest commit moves DEFAULT_REPORTING_INTERVAL_MS  into Stats.

kirktrue reviewed (commented):
Thanks for the PR @matt-welch!

I only have questions as I've not yet looked at this code 🤷‍♂️

Thanks!

matt-welch committed:
Merge branch 'trunk' into KIP-1052-PR

matt-welch committed:
Refactor inner callback conditional

matt-welch committed:
Move DEFAULT_REPORTING_INTERVAL_MS into Stats

- Remove DEFAULT_REPORTING_INTERVAL_MS as a constructor parameter and
  move into Stats class; refactor Stats constructors to not use
  reportingInterval as a parameter
- Remove DEBUG print statement

matt-welch commented:
My latest commit moves DEFAULT_REPORTING_INTERVAL_MS  into Stats and removes the DEBUG print about config.warmupRecords.  Please let me know if there are any additional changes needed on this patch. Thanks @chia7712 @kirktrue !

matt-welch commented:
@chia7712 @kirktrue I think I've addressed the issues raised, but does anyone have any more input on this patch?

matt-welch commented:
Hi @chia7712 @kirktrue .  
Does anyone have additional feedback with this PR? There doesn't appear to be any updates needed to docs/ since that archive of documentation doesn't mention the producer perf test as far as I could tell. Are there any other items that need updating? Thanks!

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -94,7 +98,11 @@ void start(String[] args) throws IOException {
                 long sendStartMs = System.currentTimeMillis();
-                cb = new PerfCallback(sendStartMs, payload.length, stats);
+                if ( config.warmupRecords > 0 && i == config.warmupRecords ) {
+                    steadyStateStats = new Stats(config.numRecords - config.warmupRecords, config.warmupRecords > 0);
```
According to https://github.com/apache/kafka/pull/17340/files#r1972675533, `config.warmupRecords > 0` could be replaced by `true`, right?

matt-welch replied:
Good point.  That second evaluation of the conditional 'config.warmupRecords > 0' is replaced by 'true' in my latest commit.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -471,6 +513,10 @@ public void onCompletion(RecordMetadata metadata, Exception exception) {
                 this.stats.iteration++;
+                if (steadyStateStats != null) {
+                    this.steadyStateStats.record(latency, bytes, now);
+                    this.steadyStateStats.iteration++;
```
Perhaps `steadyStateStats.iteration` could be updated within `steadyStateStats.record`? Since record already updates many other values, adding iteration there should be straightforward.

matt-welch replied:
Done.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -94,7 +98,11 @@ void start(String[] args) throws IOException {
-                cb = new PerfCallback(sendStartMs, payload.length, stats);
+                if ( config.warmupRecords > 0 && i == config.warmupRecords ) {
+                    steadyStateStats = new Stats(config.numRecords - config.warmupRecords, config.warmupRecords > 0);
+                    stats.steadyStateActive = true;
```
It seems this is used to stop the `printWindow`, right? If so, maybe the better naming is `stopPrint`. WDYT?

matt-welch replied:
I like this because it makes the switchover process much more clear! Since the warmup stats object continues to accumulate data during the remainder of the test, but disables only its own printing, I named the variable suppressPrint and added a method for disabling printing of the warmup stats once we've switched to steady state.

chia7712 reviewed (commented):
@matt-welch sorry for late reviews. A couple of comments are left. PTAL

matt-welch committed:
Merge branch 'trunk' of github.com:apache/kafka into KIP-1052-PR

matt-welch committed:
Refactor printing and iteration vars

- Move increment of iteration into Stats.record()
- Refactor isSteadyState to suppressPrint, add method to disable printing
  for the warmup stats object
- Remove Stats default constructor
- Remove redundant check of config.warmupRecords, replace with true

matt-welch commented:
Hi @chia7712.  I've updated my latest commit with your suggestions and I think it's a bit more readable now.  
Let me know if anyone has any more feedback.

matt-welch commented:
Hi again @chia7712 :) 
Just checking in to see if there are any more issues with this patch.  Please let me know if there's anything I can do to improve it.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -351,8 +378,10 @@ static class Stats {
         private long windowBytes;
         private long windowStart;
+        private final boolean isSteadyState;
+        private boolean supressPrint;
```
`supressPrint` -> `suppressPrint`

matt-welch replied:
Thanks for finding this one!  Fixed.

chia7712 reviewed (commented):
@matt-welch sorry for late review. one last comment is left. PTAL

matt-welch committed:
Fix typo supress->suppress

chia7712 commented:
@matt-welch thanks for your contribution and patience.

matt-welch commented:
Thank you @fvaleri @kirktrue @chia7712 for all your help and support getting this approved!
